### PR TITLE
perf: reuse allocated buffer. add config parameter

### PR DESCRIPTION
**Pull request by LiquidTheDangerous**
Added parameter -buffer-size. Reusing pre-allocated buffer for connection handling.

### CODE REVIEW

File: proxy/io.go [hunk 1]
```diff
@@ -25,45 +23,35 @@ func WriteChunks(conn *net.TCPConn, c [][]byte) (n int, err error) {
...
-		ret = append(ret, buf[:n]...)
+func ReadBytes(conn *net.TCPConn, dest []byte) ([]byte, error) {
+	n, err := readBytesInternal(conn, dest)
+	return dest[:n], err
+}
 
-		if n < BUF_SIZE {
-			break
+func readBytesInternal(conn *net.TCPConn, dest []byte) (int, error) {
+	totalRead, err := conn.Read(dest)
```
Comment (LiquidTheDangerous) [inline proxy/io.go:32]: Previously, this loop would run once, except when the entire buffer was read. Decided to remove the unnecessary loop.

File: proxy/io.go [hunk 2]
```diff
@@ -25,45 +23,35 @@ func WriteChunks(conn *net.TCPConn, c [][]byte) (n int, err error) {
 	return total, nil
 }
 
-func ReadBytes(conn *net.TCPConn) ([]byte, error) {
-	ret := make([]byte, 0)
-	buf := make([]byte, BUF_SIZE)
-
-	for {
```
Comment (LiquidTheDangerous) [inline proxy/io.go:32]: Previously, this read loop would run once, except when the entire buffer was read. Decided to remove the unnecessary loop.

File: packet/http.go [hunk 3]
```diff
@@ -56,12 +57,16 @@ func ParseUrl(raw []byte) {
 
 }
 
-func NewHttpPacket(raw []byte) (*HttpPacket, error) {
-	pkt := &HttpPacket{raw: raw}
-
-	pkt.parse()
-
-	return pkt, nil
+func NewHttpPacketFromReader(rdr io.Reader) (*HttpPacket, error) {
```
Comment (LiquidTheDangerous) [inline packet/http.go:60]: Previously, the request was parsed from a pre-read buffer, the size of which is 0 <= n < 1024. There could have been a problem if the request was sent in two parts. A reader was created, based on the underlying connection.

**Comment by xvzc** [comment]
Hi, thanks for your contribution. however, i don't know if we really need to add another option to set the buffer size.

**Comment by LiquidTheDangerous** [comment]
We can not add the buffer-size parameter to the args. How about reworking the socket reading function. In this branch, an improvement was made to reuse the allocated buffer within a single connection

File: proxy/io.go [hunk 4]
```diff
@@ -25,45 +23,35 @@ func WriteChunks(conn *net.TCPConn, c [][]byte) (n int, err error) {
...
-      )
-    }
+		if timeout > 0 {
+			from.SetReadDeadline(
+				time.Now().Add(time.Millisecond * time.Duration(timeout)),
+			)
+		}
 
-		buf, err := ReadBytes(from)
+		bytesRead, err := ReadBytes(from, buf)
```
Comment (LiquidTheDangerous) [inline proxy/io.go:54]: Here, the previously allocated buffer is reused.

**Comment by xvzc** [comment]
Yeah it's all good but we can just set the buffer-size to 1024 and remove the command line option

**Pull request merged**